ssd1306/ssd1306_main.c

```diff
@@ -0,0 +1,381 @@
+
+#define SSD1306_DRIVER_ID                           "gl_ssd1306"
+
+struct ssd1306_ctx {
+	struct work_struct display_ws;
+	struct i2c_client *client;
+	struct fb_info *info;
+	u8 *vmem;
+};
+
+static struct fb_fix_screeninfo ssd1306_fb_fix_screeninfo = {
+	.id          = "Solomon SSD1306",
+	.type        = FB_TYPE_PACKED_PIXELS,
+	.visual      = FB_VISUAL_MONO10,
+	.xpanstep    = 0,
+	.ypanstep    = 0,
+	.ywrapstep   = 0,
+	.accel       = FB_ACCEL_NONE,
+	.line_length = SSD1306_DISPLAY_WIDTH / 8,
+};
+
+static struct fb_var_screeninfo ssd1306_fb_var_screeninfo = {
+	.xres = SSD1306_DISPLAY_WIDTH,
+	.yres = SSD1306_DISPLAY_HEIGHT,
+
+	.xres_virtual = SSD1306_DISPLAY_WIDTH,
+	.yres_virtual = SSD1306_DISPLAY_HEIGHT,
+
+	.red.length   = 1,
+	.blue.length  = 1,
+	.green.length = 1,
+
+	.bits_per_pixel = 1,
+};
+
+static inline int ssd1306_write_cmd_byte(struct ssd1306_ctx *ctx, u32 cmd)
+{
+	return i2c_smbus_write_byte_data(ctx->client, SSD1306_WRITE_CMD, cmd);
+}
+
+static inline int ssd1306_write_buf(struct ssd1306_ctx *ctx, u8 *buf, size_t len)
+{
+	return i2c_master_send(ctx->client, buf, len);
+}
+
+static int ssd1306_display_on(struct ssd1306_ctx *ctx)
+{
+	ssd1306_write_cmd_byte(ctx, SSD1306_CMD_CHARGE_PUMP_SETTING);
+	ssd1306_write_cmd_byte(ctx, SSD1306_CMD_CHARGE_ENABLE);
+	ssd1306_write_cmd_byte(ctx, SSD1306_CMD_DISPLAY_ON);
+	return 0;
+}
+
+static int ssd1306_display_off(struct ssd1306_ctx *ctx)
+{
+	ssd1306_write_cmd_byte(ctx, SSD1306_CMD_CHARGE_PUMP_SETTING);
+	ssd1306_write_cmd_byte(ctx, 0x10); /* ??? */
+	ssd1306_write_cmd_byte(ctx, SSD1306_CMD_DISPLAY_OFF);
+	return 0;
+}
+
+static int ssd1306_blank(int blank_mode, struct fb_info *info)
+{
+	struct ssd1306_ctx *ctx = info->par;
+
+	if (blank_mode != FB_BLANK_UNBLANK)
+		return ssd1306_display_off(ctx);
+	else
+		return ssd1306_display_on(ctx);
+}
+
+static void ssd1306_display_update_direct(struct ssd1306_ctx *ctx)
+{
+	u8 *vmem = ctx->info->screen_base;
+	u8 *buf, *ptr;
+	int i, j, k;
+	const size_t buf_len = (SSD1306_DISPLAY_WIDTH * SSD1306_DISPLAY_HEIGHT / 8) + 1;
+
+	ptr = buf = kzalloc(buf_len, GFP_KERNEL);
+	if (!buf)
+		return;
+
+	/* Set command type */
+	buf[0] = SSD1306_DATA;
+	buf++;
+
+	for (i = 0; i < (SSD1306_DISPLAY_HEIGHT / 8); i++) {
+		for (j = 0; j < SSD1306_DISPLAY_WIDTH; j++) {
+			u32 pos = i * SSD1306_DISPLAY_WIDTH + j;
+
+			buf[pos] = 0;
+			for (k = 0; k < 8; k++) {
+				u32 page_length = SSD1306_DISPLAY_WIDTH * i;
+				u32 index = page_length + (SSD1306_DISPLAY_WIDTH * k + j) / 8;
+				u8 byte = *(vmem + index);
+				u8 bit = byte & (1 << (j % 8));
+
+				bit = bit >> (j % 8);
+				buf[pos] |= bit << k;
+			}
+		}
+	}
+
+	ssd1306_write_buf(ctx, ptr, buf_len);
+	kfree(ptr);
+}
+
+static void ssd1306_display_update_work(struct work_struct *work)
+{
+	struct ssd1306_ctx *ctx = container_of(work, struct ssd1306_ctx, display_ws);
+
+	ssd1306_display_update_direct(ctx);
+}
+
+static void ssd1306_display_update(struct ssd1306_ctx *ctx)
```

Comment:
There are several *_display_update_* methods - that's really confusing. Could you rename these?